content/posts/2019-11-07-deterministic-builds-with-clang-and-lld.md

```diff
@@ -0,0 +1,166 @@
+
+## Getting debuggers to work well with locally deterministic builds
+
+At the moment, no debugger offers an option to resolve relative paths in debug info against the directory the debugged binary is in.
+
+Some debuggers (gdb, lldb) do try to resolve relative paths against the cwd, so a simple way to make debugging work is to cd into your build directory before debugging.
+
+If you don’t want to require devs to cd into the build directory for debugging to work, you have to do debugger-specific configuration tweaks.
+
+To make sure devs don’t miss this, you could have your custom init script set an env var and query if it’s set early during your test binary startup, and exit with a message like “Add `source /path/to/your/project/gdbinit` to your ~/.gdbinit” if the environment variable isn’t set.
+
+### gdb
+
+`dir path/to/build/dir` tells gdb what directory to resolve relative paths against.
+
+`show debug-file-directory` prints the list of directories gdb looks in for dwo files. Query that, append `:path/to/build/dir`, and call `set debug-file-directory` to add your build dir to that search path.
+
+For an example, see [Chromium’s gdbinit](https://chromium.googlesource.com/chromium/src/+/master/tools/gdb/gdbinit#78) (which also does a few other unrelated things).
+
+### lldb
+
+`settings set target.source-map ../.. /absolute/path/to/build/dir` can map the “../..” prefix that all .cc files will refer to when using the setup described above with an absolute path. This requires Xcode 10.3 or newer; the lldb shipping with Xcode 10.1 has problems with this setup.
+
+For an example, see [Chromium’s lldbinit](https://chromium.googlesource.com/chromium/src/+/master/tools/lldb/lldbinit.py).
+
+### Visual Studio’s debugger and windbg
+
+If you use the setup described above, `/PDBSourcePath:X:\fake\prefix` will combine with the `..\..\my\file.cc` relative paths to make your code appear at `X:\my\file.cc`. To make Windows debuggers find them, you have two options:
+
+1. Run `subst X: C:\src\real\root` in cmd.exe before launching the debuggers to create a virtual drive that maps X: to the actual source location. Both windbg and Visual Studio will load code over X: this way.
+2. Add `C:\src\real\root` to each debugger’s source search path.
+3. - [Windbg](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/source-window#source-path): Run `.srcpath+ C:\src\real\root`. You can also set this via the `_NT_SOURCE_PATH`  environment variable, or via  File->Source File Path (Ctrl+P). Or pass `-srcpath C:\src\real\root` when launching windbg from the command line.
```

Comment:
This should be a `3.`, it's a sub-list of `2.`

Reply:
Will fix, this is one thing the converter gets wrong, then I got it wrong manually fixing it.

Reply:
Done.